scapy/contrib/automotive/someip_sd.py

```diff
@@ -0,0 +1,342 @@
+    # - key : field_name, value : desired value
+    # - it will be used from 'init_fields' function, upon packet initialization
+    #
+    # example : _defaults =
+    #  {'field_1_name':field_1_value,'field_2_name':field_2_value}
+    _defaults = {}
+
+    def _set_defaults(self):
+        for key in self._defaults.keys():
+            try:
+                self.get_field(key)
+            except KeyError:
+                pass
+            else:
+                self.setfieldval(key, self._defaults[key])
+
+    def init_fields(self):
+        Packet.init_fields(self)
+        self._set_defaults()
+
+
+# SD ENTRY
+#  - Service
+#  - EventGroup
+class _SDEntry(_SDPacketBase):
+    TYPE_FMT = ">B"
+    TYPE_PAYLOAD_I = 0
+    TYPE_SRV_FINDSERVICE = 0x00
+    TYPE_SRV_OFFERSERVICE = 0x01
+    TYPE_SRV = (TYPE_SRV_FINDSERVICE, TYPE_SRV_OFFERSERVICE)
+    TYPE_EVTGRP_SUBSCRIBE = 0x06
+    TYPE_EVTGRP_SUBSCRIBE_ACK = 0x07
+    TYPE_EVTGRP = (TYPE_EVTGRP_SUBSCRIBE, TYPE_EVTGRP_SUBSCRIBE_ACK)
+    OVERALL_LEN = 16
+
+    fields_desc = [
+        ByteField("type", 0),
+        ByteField("index_1", 0),
+        ByteField("index_2", 0),
+        BitField("n_opt_1", 0, 4),
+        BitField("n_opt_2", 0, 4),
+        ShortField("srv_id", 0),
+        ShortField("inst_id", 0),
+        ByteField("major_ver", 0),
+        X3BytesField("ttl", 0)]
+
+    def guess_payload_class(self, payload):
+        pl_type = orb(payload[_SDEntry.TYPE_PAYLOAD_I])
+
+        if (pl_type in _SDEntry.TYPE_SRV):
+            return (SDEntry_Service)
+        elif (pl_type in _SDEntry.TYPE_EVTGRP):
+            return (SDEntry_EventGroup)
+
+
+class SDEntry_Service(_SDEntry):
+    _defaults = {"type": _SDEntry.TYPE_SRV_FINDSERVICE}
+
+    name = "Service Entry"
+    fields_desc = [
+        _SDEntry,
+        IntField("minor_ver", 0)]
+
+
+class SDEntry_EventGroup(_SDEntry):
+    _defaults = {"type": _SDEntry.TYPE_EVTGRP_SUBSCRIBE}
+
+    name = "Eventgroup Entry"
+    fields_desc = [
+        _SDEntry,
+        BitField("res", 0, 12),
+        BitField("cnt", 0, 4),
+        ShortField("eventgroup_id", 0)]
+
+
+# SD Option
+#  - Configuration
+#  - LoadBalancing
+# - IPv4 EndPoint
+# - IPv6 EndPoint
+
+# - IPv4 MultiCast
+# - IPv6 MultiCast
+# - IPv4 EndPoint
+# - IPv6 EndPoint
+class _SDOption(_SDPacketBase):
+    CFG_TYPE = 0x01
+    CFG_OVERALL_LEN = 4
+    LOADBALANCE_TYPE = 0x02
+    LOADBALANCE_LEN = 0x05
+    LOADBALANCE_OVERALL_LEN = 8
+    IP4_ENDPOINT_TYPE = 0x04
+    IP4_ENDPOINT_LEN = 0x0009
+    IP4_MCAST_TYPE = 0x14
+    IP4_MCAST_LEN = 0x0009
+    IP4_SDENDPOINT_TYPE = 0x24
+    IP4_SDENDPOINT_LEN = 0x0009
+    IP4_OVERALL_LEN = 12
+    IP6_ENDPOINT_TYPE = 0x06
+    IP6_ENDPOINT_LEN = 0x0015
+    IP6_MCAST_TYPE = 0x16
+    IP6_MCAST_LEN = 0x0015
+    IP6_SDENDPOINT_TYPE = 0x26
+    IP6_SDENDPOINT_LEN = 0x0015
+    IP6_OVERALL_LEN = 24
+
+    def guess_payload_class(self, payload):
+        pl_type = orb(payload[2])
+
+        if (pl_type == _SDOption.CFG_TYPE):
```

Comment:
Avoid long `if ... elif ... elif` statements. Here, a `dict.get()` would be perfect. Also, you need a default (I have used `Raw`, but maybe something else would be better?):

```Python
        return {
            _SDOption.CFG_TYPE: SDOption_Config,
            self.LOADBALANCE_TYPE: SDOption_LoadBalance,
            [...]
        }.get(pl_type, Raw)
```

Reply:
Yes that's better, thanks for the suggestion. Raw is correct.